src/main/java/frc/robot/subsystems/drive/Drive.java

```diff
@@ -1,74 +1,285 @@
+            new Translation2d(-1.0, -1.0)
+    );
+    private final SwerveDrivePoseEstimator poseEstimator = new SwerveDrivePoseEstimator(kinematics,
+            new Rotation2d(),
+            new SwerveModulePosition[]{
+                    new SwerveModulePosition(0, new Rotation2d()),
+                    new SwerveModulePosition(0, new Rotation2d()),
+                    new SwerveModulePosition(0, new Rotation2d()),
+                    new SwerveModulePosition(0, new Rotation2d())},
+            new Pose2d());
+    private final double[] offset = new double[]{-4.414796707534875,
+            -5.833728936329093,
+            -5.522330836388308,
+            -5.927301764390117};
+    PhotonPoseEstimator photonPoseEstimator = new PhotonPoseEstimator(
+            AprilTagFields.k2023ChargedUp.loadAprilTagLayoutField(),
+            PhotonPoseEstimator.PoseStrategy.MULTI_TAG_PNP,
+            alterCam,
+            new Transform3d(
+                    new Translation3d(-0.258, 0.0, 0.137),
+                    new Rotation3d(
+                            0.0,
+                            Units.degreesToRadians(-12.0),
+                            Units.degreesToRadians(180.0))));
+    boolean testMode = false;
+    double testTime = 0.0;
+    List<ProfiledPIDController> anglePIDs;
+    SimpleMotorFeedforward angleFF;
+    List<ProfiledPIDController> drivePIDs;
+    SimpleMotorFeedforward driveFF;
+    double lastTime = Timer.getFPGATimestamp();
+
+    /**
+     * Creates a new Drive.
+     *
+     * @param flModuleIO ModuleIO for the front left swerve module.
+     * @param frModuleIO ModuleIO for the front right swerve module.
+     * @param blModuleIO ModuleIO for the back left swerve module.
+     * @param brModuleIO ModuleIO for the back right swerve module.
+     * @param gyroIO     GyroIO for gyroscope data
+     * @param visionIO   VisionIO for vision data
+     */
+    public Drive(ModuleIO flModuleIO,
+                 ModuleIO frModuleIO,
+                 ModuleIO blModuleIO,
+                 ModuleIO brModuleIO,
+                 GyroIO gyroIO,
+                 VisionIO visionIO) throws IOException {
+        this.gyroIO = gyroIO;
+        this.visionIO = visionIO;
+        moduleIOs[0] = flModuleIO;
+        moduleIOs[1] = frModuleIO;
+        moduleIOs[2] = blModuleIO;
+        moduleIOs[3] = brModuleIO;
+
+        anglePIDs = new ArrayList<>(4);
+        for (int i = 0; i < 4; i++) {
+            anglePIDs.add(new ProfiledPIDController(2.0,//4.1487,
+                    0.0,
+                    0.0,//.14002,
+                    new TrapezoidProfile.Constraints(PI * 4,
+                            PI * 8)));
+            anglePIDs.get(i).enableContinuousInput(-PI, PI);
+        }
+
+
+        angleFF = new SimpleMotorFeedforward(.25928, .28217, .0050137);
+
+
+        drivePIDs = new ArrayList<>(4);
+        for (int i = 0; i < 4; i++) {
+            drivePIDs.add(new ProfiledPIDController(
+                    2.37,
```

Comment:
shouldnt the proportion be less than that?

Reply:
Still tuning pis

Reply:
PID